map/chart_generator.cpp

```diff
@@ -0,0 +1,236 @@
+#include "map/chart_generator.hpp"
+
+#include "base/assert.hpp"
+#include "base/math.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/fstream.hpp"
+
+#include "3party/agg/agg_conv_curve.h"
+#include "3party/agg/agg_conv_stroke.h"
+#include "3party/agg/agg_path_storage.h"
+#include "3party/agg/agg_pixfmt_rgba.h"
+#include "3party/agg/agg_rasterizer_scanline_aa.h"
+#include "3party/agg/agg_renderer_primitives.h"
+#include "3party/agg/agg_renderer_scanline.h"
+#include "3party/agg/agg_scanline_p.h"
+
+namespace
+{
+template <class Color, class Order>
+struct BlendAdaptor
+{
+  using order_type = Order;
+  using color_type = Color;
+  using TValueType = typename color_type::value_type;
+  using TCalcType = typename color_type::calc_type;
+
+  enum
+  {
+    ScaleShift = color_type::base_shift,
+    ScaleMask = color_type::base_mask,
+  };
+
+  static AGG_INLINE void blend_pix(unsigned op, TValueType * p, unsigned cr, unsigned cg,
+                                   unsigned cb, unsigned ca, unsigned cover)
+  {
+    using TBlendTable = agg::comp_op_table_rgba<Color, Order>;
+    if (p[Order::A])
+    {
+      TBlendTable::g_comp_op_func[op](p, (cr * ca + ScaleMask) >> ScaleShift,
+                                      (cg * ca + ScaleMask) >> ScaleShift,
+                                      (cb * ca + ScaleMask) >> ScaleShift, ca, cover);
+    }
+    else
+    {
+      TBlendTable::g_comp_op_func[op](p, cr, cg, cb, ca, cover);
+    }
+  }
+};
+}  // namespace
+
+namespace maps
+{
+bool NormalizeChartData(vector<double> const & distanceDataM,
+                        feature::TAltitudes const & altitudeDataM, size_t resultPointCount,
+                        vector<double> & uniformAltitudeDataM)
+{
+  double constexpr kEpsilon = 1e-6;
+
+  if (distanceDataM.empty() || resultPointCount == 0 || resultPointCount == 1 ||
```

Comment:
As we discussed, it's not a problem that resultPointCount is zero or one, and the method should return true. Note that by the method contract in the absence of errors the size of uniformAltitudeDataM should be equal to resultPointCount.